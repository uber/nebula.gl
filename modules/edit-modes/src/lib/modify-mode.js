// @flow

import nearestPointOnLine from '@turf/nearest-point-on-line';
import { point, lineString as toLineString } from '@turf/helpers';
import {
  recursivelyTraverseNestedArrays,
  nearestPointOnProjectedLine,
  type NearestPointType
} from '../utils.js';
import type {
  Position,
  LineString,
  Point,
  FeatureCollection,
  FeatureOf
} from '../geojson-types.js';
import type {
  ModeProps,
  ClickEvent,
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent,
  Viewport
} from '../types.js';
import {
  BaseGeoJsonEditMode,
  getPickedEditHandle,
  getPickedEditHandles,
  getPickedExistingEditHandle,
  getPickedIntermediateEditHandle,
  getEditHandlesForGeometry,
  type GeoJsonEditAction,
  type EditHandle
} from './geojson-edit-mode.js';
import { ImmutableFeatureCollection } from './immutable-feature-collection.js';

export class ModifyMode extends BaseGeoJsonEditMode {
  getEditHandlesAdapter(
    picks: ?Array<Object>,
    mapCoords: ?Position,
    props: ModeProps<FeatureCollection>
  ): EditHandle[] {
    let handles = [];
    const { features } = props.data;

    for (const index of this.getSelectedFeatureIndexes(props)) {
      if (index < features.length) {
        const { geometry } = features[index];
        handles.push(...getEditHandlesForGeometry(geometry, index));
      } else {
        console.warn(`selectedFeatureIndexes out of range ${index}`); // eslint-disable-line no-console,no-undef
      }
    }

    // intermediate edit handle
    if (picks && picks.length && mapCoords) {
      const existingEditHandle = getPickedExistingEditHandle(picks);
      // don't show intermediate point when too close to an existing edit handle
      const featureAsPick = !existingEditHandle && picks.find(pick => !pick.isGuide);

      // is the feature in the pick selected
      if (
        featureAsPick &&
        !featureAsPick.object.geometry.type.includes('Point') &&
        this.getSelectedFeatureIndexes(props).includes(featureAsPick.index)
      ) {
        let intermediatePoint: ?NearestPointType = null;
        let positionIndexPrefix = [];
        const referencePoint = point(mapCoords);
        // process all lines of the (single) feature
        recursivelyTraverseNestedArrays(
          featureAsPick.object.geometry.coordinates,
          [],
          (lineString, prefix) => {
            const lineStringFeature = toLineString(lineString);
            const candidateIntermediatePoint = this.nearestPointOnLine(
              lineStringFeature,
              referencePoint,
              props.modeConfig && props.modeConfig.viewport
            );
            if (
              !intermediatePoint ||
              candidateIntermediatePoint.properties.dist < intermediatePoint.properties.dist
            ) {
              intermediatePoint = candidateIntermediatePoint;
              positionIndexPrefix = prefix;
            }
          }
        );
        // tack on the lone intermediate point to the set of handles
        if (intermediatePoint) {
          const {
            geometry: { coordinates: position },
            properties: { index }
          } = intermediatePoint;
          handles = [
            ...handles,
            {
              position,
              positionIndexes: [...positionIndexPrefix, index + 1],
              featureIndex: featureAsPick.index,
              type: 'intermediate'
            }
          ];
        }
      }
    }

    return handles;
  }

  // turf.js does not support elevation for nearestPointOnLine
  nearestPointOnLine(
    line: FeatureOf<LineString>,
    inPoint: FeatureOf<Point>,
    viewport: ?Viewport
  ): NearestPointType {
    const { coordinates } = line.geometry;
    if (coordinates.some(coord => coord.length > 2)) {
      if (viewport) {
        // This line has elevation, we need to use alternative algorithm
        return nearestPointOnProjectedLine(line, inPoint, viewport);
      }
      // eslint-disable-next-line no-console,no-undef
      console.log(
        'Editing 3D point but modeConfig.viewport not provided. Falling back to 2D logic.'
      );
    }

    return nearestPointOnLine(line, inPoint);
  }

  handleClickAdapter(event: ClickEvent, props: ModeProps<FeatureCollection>): ?GeoJsonEditAction {
    let editAction: ?GeoJsonEditAction = null;

    const pickedExistingHandle = getPickedExistingEditHandle(event.picks);
    const pickedIntermediateHandle = getPickedIntermediateEditHandle(event.picks);

    if (pickedExistingHandle) {
      let updatedData;
      try {
        updatedData = new ImmutableFeatureCollection(props.data)
          .removePosition(pickedExistingHandle.featureIndex, pickedExistingHandle.positionIndexes)
          .getObject();
      } catch (ignored) {
        // This happens if user attempts to remove the last point
      }

      if (updatedData) {
        editAction = {
          updatedData,
          editType: 'removePosition',
          editContext: {
            featureIndexes: [pickedExistingHandle.featureIndex],
            positionIndexes: pickedExistingHandle.positionIndexes,
            position: pickedExistingHandle.position
          }
        };
      }
    } else if (pickedIntermediateHandle) {
      const updatedData = new ImmutableFeatureCollection(props.data)
        .addPosition(
          pickedIntermediateHandle.featureIndex,
          pickedIntermediateHandle.positionIndexes,
          pickedIntermediateHandle.position
        )
        .getObject();

      if (updatedData) {
        editAction = {
          updatedData,
          editType: 'addPosition',
          editContext: {
            featureIndexes: [pickedIntermediateHandle.featureIndex],
            positionIndexes: pickedIntermediateHandle.positionIndexes,
            position: pickedIntermediateHandle.position
          }
        };
      }
    }
    return editAction;
  }

  handlePointerMove(event: PointerMoveEvent, props: ModeProps<FeatureCollection>): void {
    let editAction: ?GeoJsonEditAction = null;

    const editHandle = getPickedEditHandle(event.pointerDownPicks);

    if (event.isDragging && editHandle) {
      const updatedData = new ImmutableFeatureCollection(props.data)
        .replacePosition(editHandle.featureIndex, editHandle.positionIndexes, event.mapCoords)
        .getObject();

      editAction = {
        updatedData,
        editType: 'movePosition',
        editContext: {
          featureIndexes: [editHandle.featureIndex],
          positionIndexes: editHandle.positionIndexes,
          position: event.mapCoords
        }
      };

      props.onEdit(editAction);
    }

    const cursor = this.getCursor(event);
    props.onUpdateCursor(cursor);

    // Cancel map panning if pointer went down on an edit handle
    const cancelMapPan = Boolean(editHandle);
    if (cancelMapPan) {
      event.sourceEvent.stopPropagation();
    }
  }

  handleStartDraggingAdapter(
    event: StartDraggingEvent,
    props: ModeProps<FeatureCollection>
  ): ?GeoJsonEditAction {
    let editAction: ?GeoJsonEditAction = null;

    const selectedFeatureIndexes = this.getSelectedFeatureIndexes(props);

    const editHandle = getPickedIntermediateEditHandle(event.picks);
    if (selectedFeatureIndexes.length && editHandle) {
      const updatedData = new ImmutableFeatureCollection(props.data)
        .addPosition(editHandle.featureIndex, editHandle.positionIndexes, event.mapCoords)
        .getObject();

      editAction = {
        updatedData,
        editType: 'addPosition',
        editContext: {
          featureIndexes: [editHandle.featureIndex],
          positionIndexes: editHandle.positionIndexes,
          position: event.mapCoords
        }
      };
    }

    return editAction;
  }

  handleStopDraggingAdapter(
    event: StopDraggingEvent,
    props: ModeProps<FeatureCollection>
  ): ?GeoJsonEditAction {
    let editAction: ?GeoJsonEditAction = null;

    const selectedFeatureIndexes = this.getSelectedFeatureIndexes(props);
    const editHandle = getPickedEditHandle(event.picks);
    if (selectedFeatureIndexes.length && editHandle) {
      const updatedData = new ImmutableFeatureCollection(props.data)
        .replacePosition(editHandle.featureIndex, editHandle.positionIndexes, event.mapCoords)
        .getObject();

      editAction = {
        updatedData,
        editType: 'finishMovePosition',
        editContext: {
          featureIndexes: [editHandle.featureIndex],
          positionIndexes: editHandle.positionIndexes,
          position: event.mapCoords
        }
      };
    }

    return editAction;
  }

  getCursor(event: PointerMoveEvent): ?string {
    const picks = (event && event.picks) || [];

    const handlesPicked = getPickedEditHandles(picks);
    if (handlesPicked.length) {
      return 'cell';
    }
    return null;
  }
}
