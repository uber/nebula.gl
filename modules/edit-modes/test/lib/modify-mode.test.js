// @flow
/* eslint-env jest */

import { ModifyMode } from '../../src/lib/modify-mode.js';
import { createFeatureCollectionProps } from '../test-utils.js';

let pointFeature;
let lineStringFeature;
let polygonFeature;
let multiPointFeature;
let multiLineStringFeature;
let multiPolygonFeature;

beforeEach(() => {
  pointFeature = {
    type: 'Feature',
    properties: {},
    geometry: { type: 'Point', coordinates: [1, 2] },
  };

  lineStringFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'LineString',
      coordinates: [
        [1, 2],
        [2, 3],
        [3, 4],
      ],
    },
  };

  polygonFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'Polygon',
      coordinates: [
        // exterior ring
        [
          [-1, -1],
          [1, -1],
          [1, 1],
          [-1, 1],
          [-1, -1],
        ],
        // hole
        [
          [-0.5, -0.5],
          [-0.5, 0.5],
          [0.5, 0.5],
          [0.5, -0.5],
          [-0.5, -0.5],
        ],
      ],
    },
  };

  multiPointFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'MultiPoint',
      coordinates: [
        [1, 2],
        [3, 4],
      ],
    },
  };

  multiLineStringFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'MultiLineString',
      coordinates: [
        [
          [1, 2],
          [2, 3],
          [3, 4],
        ],
        [
          [5, 6],
          [6, 7],
          [7, 8],
        ],
      ],
    },
  };

  multiPolygonFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'MultiPolygon',
      coordinates: [
        [
          // exterior ring polygon 1
          [
            [-1, -1],
            [1, -1],
            [1, 1],
            [-1, 1],
            [-1, -1],
          ],
          // hole  polygon 1
          [
            [-0.5, -0.5],
            [-0.5, 0.5],
            [0.5, 0.5],
            [0.5, -0.5],
            [-0.5, -0.5],
          ],
        ],
        [
          // exterior ring polygon 2
          [
            [2, -1],
            [4, -1],
            [4, 1],
            [2, 1],
            [2, -1],
          ],
        ],
      ],
    },
  };
});

describe('getGuides()', () => {
  it('gets edit handles for Point', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [pointFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for LineString', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineStringFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for Polygon', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [polygonFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for MultiPoint', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [multiPointFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for MultiLineString', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [multiLineStringFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for MultiPolygon', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [multiPolygonFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for all selected features in collection', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineStringFeature, pointFeature, multiPointFeature],
      },
      selectedIndexes: [0, 2],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  const lineString = {
    type: 'Feature',
    geometry: {
      type: 'LineString',
      coordinates: [
        [-122.40880966186523, 37.783536601521924],
        [-122.43893623352051, 37.779669924659004],
        [-122.43515968322752, 37.7624370109886],
        [-122.42348670959471, 37.77180027337861],
        [-122.4250316619873, 37.778584505321376],
        [-122.42314338684082, 37.778652344496926],
        [-122.42357254028322, 37.77987343901049],
        [-122.41198539733887, 37.78109451335266],
      ],
    },
  };

  const point = {
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [-122.40880966186523, 37.783536601521924],
    },
  };
  const pick = {
    object: lineString,
    isGuide: false,
    index: 0,
  };

  const mapCoords = [-122.43862233312133, 37.77767798407437];

  it('includes an intermediate edit handle', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineString],
      },
      selectedIndexes: [0],
      lastPointerMoveEvent: {
        picks: [pick],
        mapCoords,
        screenCoords: [42, 42],
        cancelPan: jest.fn(),
        sourceEvent: null,
      },
    });

    const guides = mode.getGuides(props);

    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );

    expect(intermediate).toMatchSnapshot();
  });

  it('does not add intermeidate edit handle when no picks provided', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineString],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );
    expect(intermediate).toBeUndefined();
  });

  it('does not add intermeidate edit handle when too close to existing edit handle', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineString],
      },
      selectedIndexes: [0],
      lastPointerMoveEvent: {
        picks: [
          pick,
          {
            isGuide: true,
            index: 42,
            object: {
              properties: { guideType: 'editHandle', editHandleType: 'existing', featureIndex: 0 },
              geometry: { coordinates: [] },
            },
          },
        ],
        mapCoords,
        screenCoords: [42, 42],
        cancelPan: jest.fn(),
        sourceEvent: null,
      },
    });

    const guides = mode.getGuides(props);

    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );
    expect(intermediate).toBeUndefined();
  });

  it('does not add intermeidate edit handle when pick is not a selected feature', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineString],
      },
    });

    const guides = mode.getGuides(props);

    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );
    expect(intermediate).toBeUndefined();
  });

  it('does not add intermeidate edit handle when pick is a Point / MultiPoint', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [point],
      },
      selectedIndexes: [0],
      lastPointerMoveEvent: {
        picks: [
          {
            isGuide: false,
            index: 0,
            object: point,
          },
        ],
        mapCoords,
        screenCoords: [42, 42],
        cancelPan: jest.fn(),
        sourceEvent: null,
      },
    });
    const guides = mode.getGuides(props);
    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );
    expect(intermediate).toBeUndefined();
  });
});
