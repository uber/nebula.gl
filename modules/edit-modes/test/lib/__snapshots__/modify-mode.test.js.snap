// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGuides() gets edit handles for LineString 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          3,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          2,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for MultiLineString 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          3,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          5,
          6,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          6,
          7,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          7,
          8,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          2,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for MultiPoint 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for MultiPolygon 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
          3,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.5,
          -0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.5,
          0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.5,
          0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.5,
          -0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
          3,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          4,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          4,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
          3,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for Point 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for Polygon 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          3,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.5,
          -0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.5,
          0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.5,
          0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.5,
          -0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          3,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for all selected features in collection 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          3,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 2,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 2,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() includes an intermediate edit handle 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      -122.43850292231143,
      37.777692666558565,
    ],
    "type": "Point",
  },
  "properties": Object {
    "editHandleType": "intermediate",
    "featureIndex": 0,
    "guideType": "editHandle",
    "positionIndexes": Array [
      2,
    ],
  },
  "type": "Feature",
}
`;
